python/across_api/base/fov.py

```diff
@@ -0,0 +1,382 @@
+            return float(round(np.sum(healpix_loc[visible_probability_pixels]), 5))
+
+
+class FootprintFOV(FOVBase):
+    """
+    Constrained instrumet FOV. This is an FOV that calculate lates what is
+    visible in the FOV at a given pointing
+    """
+
+    footprint: Footprint
+
+    def __init__(self, pointing: PointingBase) -> None:
+        projected_footprint = self.footprint.project(
+            ra=pointing.ra, dec=pointing.dec, pos_angle=pointing.position_angle
+        )
+
+        ras_poly = [x[0] for x in projected_footprint][:-1]
+        decs_poly = [x[1] for x in projected_footprint][:-1]
+
+        cartesian_vertices = spherical_to_cartesian(
+            1,  # radial component
+            np.deg2rad(decs_poly),  # LAT - DECs
+            np.deg2rad(ras_poly),  # LON - RAs
+        )
+
+        self.visible_pixels = hp.query_polygon(
+            hp.order2nside(HEALPIX_MAP_EVAL_ORDER),
+            np.array(cartesian_vertices).T,
+            inclusive=True,
+            nest=True,
+        )
+
+
+class AllSkyFOV(FOVBase):
+    """
+    All sky instrument FOV. This is a simple FOV that is always visible unless
+    Earth occulted.
+    """
+
+    earth_constraint: EarthLimbConstraint
+
+    def __init__(self, ephem: EphemBase, time: Time):
+        """
+        Finds all healpix pixels within the ephemeris time slot that are not earth
+        occulted, and sets them to FOVBase.visible_pixels.
+
+        Function does assume that it is a space mission with available ephemeris
+        """
+        n_side = hp.order2nside(HEALPIX_MAP_EVAL_ORDER)
+        n_pix = hp.order2npix(HEALPIX_MAP_EVAL_ORDER)
+        i_slice = get_slice(time=time, ephem=ephem)
+
+        ra = ephem.earth.ra[i_slice].degree
+        dec = ephem.earth.dec[i_slice].degree
+
+        vec = hp.ang2vec(ra, dec, lonlat=True)
+
+        # np.array to store occulted pixels
+        earth_pixels = np.array([]).astype(int)
+
+        # loop over each time-slice vector and concatenate
+        # earth constrained pixels calculated with healpy's cone search (query_disc)
+        # assume the earth is a disc at ephem.earth position with radius
+        # radius = earthsize + constraint min angle.
+        for i, v in enumerate(vec):
+            radius = np.radians(ephem.earthsize[i].value) + np.radians(
+                self.earth_constraint.min_angle.value
+            )
+            earth_pixels = np.concatenate(
+                (
+                    earth_pixels,
+                    hp.query_disc(n_side, v, radius, nest=True, inclusive=False),
+                )
+            )
+
+        # only need unique pixels
+        earth_pixels = np.unique(earth_pixels)
+
+        # simulate the whole sky
+        all_sky_pixels = np.zeros(n_pix, dtype=np.float32)
+
+        # set the earth values on the sky
+        all_sky_pixels[earth_pixels] = 1
+
+        # find the complement (where sky != 1)
+        self.visible_pixels = np.where(all_sky_pixels == 0)[0]
```

Comment:
You can avoid this set complement by instead calling `query_disc` with an angle that is 180° minus the constraint angle. (Right?)

Reply:
In theory calculating the `hp.query_disc` with the antipodal position and a radius of `pi/2 - radius (radians)` will give you the same result. 

It shouldn't really be all that more performative since how its originally implemented is in numpy, and that is fast. 

I can implement it in a way:

```python
    def __init__(self, ephem: EphemBase, time: Time):
        """
        Finds all healpix pixels within the ephemeris time slot that are not earth
        occulted, and sets them to FOVBase.visible_pixels.

        This is done by finding the antipodal position of the earth and calculating
        all of the pixels within a disc of radius: 180 (deg) - radius_of_earth (deg)

        Antipodal:
            a_ra = ra-180 (deg)
            a_dec = -dec  (deg)
            a_radius = 180-radius (deg)

        Function does assume that it is a space mission with available ephemeris
        """
        n_side = hp.order2nside(HEALPIX_MAP_EVAL_ORDER)
        i_slice = get_slice(time=time, ephem=ephem)

        #antipodal positions
        ra = ephem.earth.ra[i_slice].degree - 180.0
        dec = -1.0 * ephem.earth.dec[i_slice].degree

        vec = hp.ang2vec(ra, dec, lonlat=True)

        # np.array to store visible pixels
        visible_pixels = np.array([]).astype(int)

        # loop over each time-slice vector and concatenate
        # visible pixels calculated with healpy's cone search (query_disc)
        # assume the earth is a disc at ephem.earth position with radius
        # radius = 180 - earthsize + constraint min angle.
        for i, v in enumerate(vec):
            radius = np.radians(180) - np.radians(ephem.earthsize[i].value) + np.radians(
                self.earth_constraint.min_angle.value
            )
            visible_pixels = np.concatenate(
                (
                    visible_pixels,
                    hp.query_disc(n_side, v, radius, nest=True, inclusive=False),
                )
            )

        # only need unique pixels
        self.visible_pixels = np.unique(visible_pixels)
```

Reply:
That passed the tests, so I'll push it since it will be more efficient.